Fix edge case in request manager

ptschip opened the pull request:
This fixes an edge condition which happens when we have just one single peer connected and we request a block when a header comes in but then we request a duplicate block when the IBD process kicks in. This is caused by the fact that we've added a duplicate source for the same peer into the request manager when the block was already in flight.

ptschip committed:
Fix the edge case where we download two dupicate blocks from same peer

This can happen when we have only one peer connected and we end up
adding a duplicate source for the same peer, once when the header
is received and then once when the IBD process kicks off. This can
result in a re-request of the same block from the same peer when the
block download timeout in the request manager is exceeded.

sickpig commented:
I know this PR is still WIP, but I though the following comments could be of some use nonetheless.

formatting error in `requestManager.cpp`

https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/655209812#L2395

then there are 2 travis job both failing at `graphene_version.py`, by looking at the output those seems spurious but I could be wrong

ptschip committed:
Use one CNode critical section to lock thintype block processing

Don't process two thintype blocks received for the same peer
at the same time. This prevents us from having two threads modify
the pblock data simultaneously.

ptschip committed:
Check if graphene version is supported before requesting a graphene block

Thanks @bissias for the code snippet.

This fixes an issue which showed up in graphene_versions.py where
we were requesting graphene block from a peer that didn't support
our version.

ptschip commented:
tests are passing, taking out of WIP
